### PR TITLE
ggml : add GGML_OP_LOG_SOFT_MAX op (CPU, Metal, CUDA)

### PR DESCRIPTION
**Pull request by lifeiteng**
## Summary

Add a dedicated `log_softmax` operator `GGML_OP_LOG_SOFT_MAX`, computed along `ne[0]`,
numerically stable and aligned with PyTorch float32 `log_softmax`:

```
out = (x - max) - log(sum(exp(x - max)))
```

Composing it as `ggml_log(ggml_soft_max(x))` is numerically poor (the `log` of small
softmax probabilities loses precision). A dedicated op keeps the stable form and matches
`torch.log_softmax` closely.

## Public API

```c
struct ggml_tensor * ggml_log_soft_max        (struct ggml_context * ctx, struct ggml_tensor * a);
struct ggml_tensor * ggml_log_soft_max_inplace(struct ggml_context * ctx, struct ggml_tensor * a);
```

## Backends

- **CPU** — FP32 accumulation (tracks torch float32 more closely than `double`; measured
  `max|diff|` vs `torch.log_softmax` ≈ `1.9e-6`). In-place safe.
- **Metal** — native kernels (scalar + vec4), mirroring `kernel_soft_max`'s parallel
  reduction. Runs on Metal with no CPU fallback.
- **CUDA** — native kernel (one block per row, `block_reduce` for max and sum).

`F32` only (asserted at op construction). Other backends fall back to CPU via the
scheduler. Backward is not implemented (guarded with a clear assert).

## Tests

- `test-backend-ops` cases for `LOG_SOFT_MAX` (out-of-place / in-place, 4D shapes,
  `ne0` multiple-of-4 and odd to exercise both vec4 and scalar paths).
- A standalone parity test `tests/test-log-soft-max` checks the CPU op against
  PyTorch float32 golden values (tolerance `5e-5`, observed worst `1.9e-6`), plus
  in-place == out-of-place, batched-row independence, and `sum(exp(out)) == 1`.

## Misc

- `docs/ops.md` regenerated for the new op.
- Bumped `RPC_PROTO_PATCH_VERSION` and all `GGML_OP_COUNT` static asserts.
- Wired into `ggml-backend-meta` split handling so multi-backend graphs don't abort.

### CODE REVIEW

**Comment by lifeiteng** [comment]
```
┌──────────────┬───────────┬───────────┬───────────┬────────────────┐
│ [ne0,nrows]  │    CPU    │   CUDA0   │  WebGPU   │ WebGPU vs CUDA │
├──────────────┼───────────┼───────────┼───────────┼────────────────┤
│ [1024,512]   │ 7.51 GB/s │ 698 GB/s  │ 275 GB/s  │ 2.5×        │
├──────────────┼───────────┼───────────┼───────────┼────────────────┤
│ [4096,512]   │ 5.61 GB/s │ 1622 GB/s │ 849 GB/s  │ 1.9×        │
├──────────────┼───────────┼───────────┼───────────┼────────────────┤
│ [16384,512]  │ 5.83 GB/s │ 1960 GB/s │ 1420 GB/s │ 1.4×        │
├──────────────┼───────────┼───────────┼───────────┼────────────────┤
│ [16384,4096] │ 8.48 GB/s │ 854 GB/s  │ 809 GB/s  │ 1.06×       │
├──────────────┼───────────┼───────────┼───────────┼────────────────┤
│ [65536,4096] │ 8.59 GB/s │ 668 GB/s  │ 648 GB/s  │ 1.03×       │
└──────────────┴───────────┴───────────┴───────────┴────────────────┘
```

**Comment by am17an** [comment]
What model is the operator used in?

**Comment by lifeiteng** [comment]
> am17an

log_softmax is a critical, numerically stable primitive essential for computing losses during the model training phase (preventing the precision loss or -inf underflow caused by chaining log(softmax)).

**Comment by am17an** [comment]
This should be called inside cross_entropy_loss and the PR should show actual benefits in a use-case (e.g. training) rather than hand-waving numerical robustness. Also please adhere to the contribution guidelines, do not club multiple backends in the same PR.
